libp2p/switch.nim

```diff
@@ -189,41 +192,40 @@ proc upgradeIncoming(s: Switch, conn: Connection) {.async, gcsafe.} =
   # handle secured connections
   await ms.handle(conn)
 
-proc dial*(s: Switch, 
-           peer: PeerInfo, 
-           proto: string = ""): 
-           Future[Connection] {.async.} = 
+proc dial*(s: Switch,
+           peer: PeerInfo,
+           proto: string = ""):
+           Future[Connection] {.async.} =
   let id = peer.id
-  trace "dialing peer", peer = id
-  for t in s.transports: # for each transport
-    for a in peer.addrs: # for each address
-      if t.handles(a): # check if it can dial it
-        if id notin s.connections:
-          trace "dialing address", address = $a
+  trace "Dialing peer", peer = id
+  result = s.connections.getOrDefault(id)
+  if result == nil or result.closed:
+    for t in s.transports: # for each transport
+      for a in peer.addrs: # for each address
+        if t.handles(a): # check if it can dial it
+          trace "Dialing address", address = $a
           result = await t.dial(a)
           # make sure to assign the peer to the connection
-          result.peerInfo = some(peer)
-        result = await s.upgradeOutgoing(result)
-        result.closeEvent.wait().addCallback(
-          proc(udata: pointer) =
+          result.peerInfo = some peer
+          result = await s.upgradeOutgoing(result)
+          result.closeEvent.wait().addCallback do (udata: pointer):
             asyncCheck s.cleanupConn(result)
-          )
-
-        if proto.len > 0 and not result.closed:
-          let stream = await s.getMuxedStream(peer)
-          if stream.isSome:
-            trace "connection is muxed, return muxed stream"
-            result = stream.get()
-            trace "attempting to select remote", proto = proto
-
-          if not (await s.ms.select(result, proto)):
-            error "unable to select protocol: ", proto = proto
-            raise newException(CatchableError, 
-            &"unable to select protocol: {proto}")
-
-        break # don't dial more than one addr on the same transport
-
-proc mount*[T: LPProtocol](s: Switch, proto: T) {.gcsafe.} = 
+        break
```

Comment:
This will end the loop prematurely - it needs to be placed within the if block right above it.

Reply:
Right, thanks :)